src/producer.rs

```diff
@@ -1,242 +1,343 @@
-                                let gcno_stem = gcno_stem.to_str().expect("Failed to create stem file string");
-                                queue.push(Some(WorkItem {
-                                    format: ItemFormat::GCNO,
-                                    item: ItemType::Buffers(GcnoBuffers {stem: gcno_stem.to_string(),
-                                                                         gcno_buf: Arc::clone(gcno_buf),
-                                                                         gcda_buf: Vec::new()}),
-                                    name: gcda_archive_name.to_string(),
-                                }));
-                            },
-                            None => {
-                                queue.push(Some(WorkItem {
-                                    format: ItemFormat::GCNO,
-                                    item: ItemType::Path(gcno_path),
-                                    name: gcda_archive_name.to_string(),
+                                    name: gcda_arch.get_name().to_string(),
                                 }));
                             }
                         }
+                    };
+                }
+            },
+            None => {
+                if !ignore_orphan_gcno {
+                    let archive = *archive;
+                    let gcno = format!("{}.gcno", stem).to_string();
+                    if is_llvm {
+                        let mut buffer: Vec<u8> = Vec::new();
+                        archive.read_in_buffer(&gcno, &mut buffer);
+
+                        queue.push(Some(WorkItem {
+                            format: ItemFormat::GCNO,
+                            item: ItemType::Buffers(GcnoBuffers {stem: gcno,
+                                                                 gcno_buf: Arc::new(buffer),
+                                                                 gcda_buf: Vec::new()}),
+                            name: archive.get_name().to_string(),
+                        }));
+                    } else {
+                        let physical_gcno_path = tmp_dir.join(format!("{}_{}.gcno", stem, 1));
+                        if archive.extract(&gcno, &physical_gcno_path) {
+                            queue.push(Some(WorkItem {
+                                format: ItemFormat::GCNO,
+                                item: ItemType::Path(physical_gcno_path),
+                                name: archive.get_name().to_string(),
+                            }));
+                        }
                     }
                 }
             }
         }
     }
+}
 
-    for &mut (archive_name, ref mut archive) in &mut info_archives {
-        for i in 0..archive.len() {
-            let mut file = archive.by_index(i).unwrap();
-
-            if file.name().ends_with('/') {
-                continue;
-            }
-
+pub fn info_producer(infos: RefCell<HashMap<String, Vec<&Archive>>>, queue: &WorkQueue) {
```

Comment:
Maybe in the signature of the function we can use `&HashMap` instead of `RefCell<HashMap`?